### PR TITLE
Add vlm support in `train_on_responses_only`

### PR DESCRIPTION
**Pull request by patel-zeel**
Edited `train_on_responses_only` function to support VLMs.

Test Colab: https://colab.research.google.com/drive/18PkC1dYETTdbGzBj6GwUW5cv3T3NmGYf?usp=sharing

Main changes:
- Vision model detection by checking the following VLM-specific keywards in names of `trainer.model.named_modules()`
    - `visual`: Qwen2-VL models
    - `vision_tower`: Llava/Pixtral (Mistral AI) models
    - `vision_model`: Llama models
- `trainer.train_dataset` is a list of messages. Remove a content entry if "role" is "user" and "type" is "text".

Input Example:
```py
{'messages': [{'role': 'user',
   'content': [{'type': 'text',
     'text': 'Write the LaTeX representation for this image.'},
    {'type': 'image',
     'image': <PIL.PngImagePlugin.PngImageFile image mode=RGB size=160x40>}]},
  {'role': 'assistant',
   'content': [{'type': 'text',
     'text': '{ \\frac { N } { M } } \\in { \\bf Z } , { \\frac { M } { P } } \\in { \\bf Z } , { \\frac { P } { Q } } \\in { \\bf Z }'}]}]}
```

Output Example:
```py
{'messages': [{'role': 'user',
   'content': [{'type': 'image',
     'image': <PIL.PngImagePlugin.PngImageFile image mode=RGB size=160x40>}]},
  {'role': 'assistant',
   'content': [{'type': 'text',
     'text': '{ \\frac { N } { M } } \\in { \\bf Z } , { \\frac { M } { P } } \\in { \\bf Z } , { \\frac { P } { Q } } \\in { \\bf Z }'}]}]}
```

### CODE REVIEW

**Comment by patel-zeel** [comment]
Sorry, I realized my solution is wrong because my conceptual understanding of "training on responses only" was wrong. Now, my updated understanding is that,
1. If no collator is passed, the default collator `DataCollatorForLanguageModeling` used in SFTTrainer creates "labels" identical to "input_ids".
2. As soon as the trainer is instantiated, data is tokenized, and `input_ids` are created.
3. `train_on_responses_only` function post-processes `trainer.train_dataset` and `trainer.eval_dataset` to mask out tokens for "instruction_part" because we want LLM to reconstruct answers correctly and ignore reconstruction loss for the questions.

I see that #60 is on the right track.
